climada/util/yearsets.py

```diff
@@ -0,0 +1,247 @@
+            The sampling_dict needs to be obtained in a first call,
+            i.e. [annual_impacts, sampling_dict] = climada_yearsets.impact_yearset(...)
+            and can then be provided in subsequent calls(s) to obtain the exact same sampling
+            (also for a different event_impacts object)
+        correction_fac : boolean
+            If True a correction factor is applied to the resulting annual_impacts. They are
+            scaled in such a way that the expected annual impact (eai) of the annual_impacts
+            equals the eai of the events_impacts
+
+    Returns:
+      annual_impacts : impact object
+          annual impacts for all sampled_years
+      sampling_dict : dict
+          the sampling dictionary containing two arrays:
+              selected_events (array) : sampled events (len: total amount of sampled events)
+              events_per_year (array) : events per sampled year
+          Can be used to re-create the exact same annual_impacts yearset
+      """
+
+
+    if not sampled_years and not sampling_dict:
+        sampled_years = np.arange(1, 1000+1).tolist()
+    elif isinstance(sampled_years, int):
+        sampled_years = np.arange(1, sampled_years+1).tolist()
+    elif not sampled_years:
+        sampled_years = np.arange(1, len(sampling_dict['selected_events'])+1).tolist()
+    elif len(sampled_years) != len(sampling_dict['events_per_year']):
+        LOGGER.info("The number of sampled_years and the length of the list of events_per_year "
+                    "in the sampling_dict differ. The number of years contained in the "
+                    "sampling_dict are used as number of sampled_years.")
+        sampled_years = np.arange(1, len(sampling_dict['selected_events'])+1).tolist()
+
+    if sampling_dict and (
+            sum(sampling_dict['events_per_year']) != len(sampling_dict['selected_events'])):
+        raise ValueError("The sampling dictionary is faulty: the sum of selected events "
+                         "does not correspond to the number of selected events.")
+
+    n_sampled_years = len(sampled_years)
+
+    if len(np.unique(event_impacts.frequency)) > 1:
+        LOGGER.warning("The frequencies of the single events in the given event_impacts "
+                       "differ among each other. Please beware that this will influence "
+                       "the resulting annual_impacts as the events are sampled uniformaly "
+                       "and different frequencies are (not yet) taken into account.")
+
+
+    # INITIALISATION
+    # to do: add a flag to distinguish annual_impacts from other impact class objects
+    # artificially generate a generic  annual impact set (by sampling the event impact set)
+    annual_impacts = copy.deepcopy(event_impacts)
+    annual_impacts.event_id = [] # to indicate not by event any more
+    annual_impacts.event_id = np.arange(1, n_sampled_years+1)
+    annual_impacts.frequency = [] # init, see below
```

Comment:
@chahank: does it make sense to have this line or is the overwriting of the frequency at a later stage sufficient?

Reply:
It is not necessary at all and adds data handling, but might increase readability. You can choose to keep or remove.